File: ios/LaneShadow/Views/Molecules/SearchBar.swift
import LaneShadowTheme
import SwiftUI

// MARK: - SearchBar Component

/**
 * SearchBar molecule component
 *
 * Simple search input with icon. Following React Native component from
 * react-native/components/ui/search-bar.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.surface.default`
 *   - Icon: `theme.colors.onSurface.default` with 0.6 opacity
 *   - Text (typed): `theme.colors.onSurface.default`
 *   - Text (placeholder): `theme.colors.onSurface.default` with 0.6 opacity
 * - Layout:
 *   - Corner radius: `theme.radius.md` (8pt)
 *   - Padding: horizontal 16pt, vertical 12pt
 *   - Gap: 12pt between icon and text
 * - Typography:
 *   - Font size: 14pt
 *   - Font weight: regular
 * - Icon:
 *   - Size: 20pt
 *   - SF Symbol: "magnifyingglass"
 *
 * ## Parameters
 * - placeholder: Placeholder text to display
 * - value: Current input value binding (nil for clickable mode, non-nil for editable)
 * - onTap: Tap handler for expanding search (clickable mode, default: nil)
 * - testID: Optional testing identifier for accessibility
 *
 * ## Modes
 * - **Clickable mode** (value = nil): Displays placeholder as non-editable text with onTap handler
 * - **Editable mode** (value != nil): Full TextField with text input binding
 */
public struct LSSearchBar: View {
    @Environment(\.theme) private var theme
    @FocusState private var isFocused: Bool

    private let placeholder: String
    private let value: Binding<String>?
    private let onTap: (() -> Void)?
    private let testID: String?

    public init(
        placeholder: String,
        value: Binding<String>? = nil,
        onTap: (() -> Void)? = nil,
        testID: String? = nil
    ) {
        self.placeholder = placeholder
        self.value = value
        self.onTap = onTap
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        Group {
            if let value {
                // Editable mode: TextField with input
                HStack(spacing: 12) {
                    Image(systemName: "magnifyingglass")
                        .font(.system(size: 20, weight: .regular))
                        .foregroundStyle(iconColor)

                    TextField(text: value) {
                        Text(placeholder)
                            .font(.system(size: 14, weight: .regular))
                            .foregroundStyle(iconColor)
                    }
                    .font(.system(size: 14, weight: .regular))
                    .foregroundStyle(theme.colors.onSurface.default)
                    .focused($isFocused)
                    .accessibilityLabel(placeholder)
                    .accessibilityIdentifier(testID.map { "\($0)-input" } ?? "search-bar-input")
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 12)
                .background(theme.colors.surface.default)
                .clipShape(RoundedRectangle(cornerRadius: theme.radius.md))
                .accessibilityElement(children: .contain)
                .accessibilityLabel("Search input")
            } else {
                // Clickable mode: Non-editable with onTap handler
                Button(action: {
                    onTap?()
                }) {
                    HStack(spacing: 12) {
                        Image(systemName: "magnifyingglass")
                            .font(.system(size: 20, weight: .regular))
                            .foregroundStyle(iconColor)

                        Text(placeholder)
                            .font(.system(size: 14, weight: .regular))
                            .foregroundStyle(iconColor)
                            .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)
                    .background(theme.colors.surface.default)
                    .clipShape(RoundedRectangle(cornerRadius: theme.radius.md))
                }
                .buttonStyle(PlainButtonStyle())
                .accessibilityLabel(placeholder)
                .accessibilityIdentifier(testID ?? "search-bar")
            }
        }
    }

    // MARK: - Private Computed Properties

    private var iconColor: Color {
        theme.colors.onSurface.default.opacity(0.6)
    }
}

// MARK: - Preview

#Preview("Clickable Mode") {
    VStack(spacing: 16) {
        LSSearchBar(
            placeholder: "Search rides...",
            onTap: { print("Search tapped") }
        )

        LSSearchBar(
            placeholder: "Search locations..."
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("Editable Mode") {
    VStack(spacing: 16) {
        LSSearchBar(
            placeholder: "Search rides...",
            value: .constant("")
        )

        LSSearchBar(
            placeholder: "With text",
            value: .constant("San Francisco")
        )

        LSSearchBar(
            placeholder: "Long text",
            value: .constant("This is a longer search query to demonstrate text wrapping")
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("States") {
    VStack(spacing: 16) {
        LSSearchBar(
            placeholder: "Empty",
            value: .constant("")
        )

        LSSearchBar(
            placeholder: "With text",
            value: .constant("Mission District")
        )

        LSSearchBar(
            placeholder: "Clickable",
            onTap: { print("Tapped") }
        )
    }
    .padding()
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/SearchBar.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.components.atoms.IconSymbol

/**
 * SearchBar molecule component
 *
 * Simple search input with icon. Following React Native wrapper patterns from
 * react-native/components/ui/search-bar.tsx
 *
 * Supports two modes:
 * - Clickable mode: Displays placeholder/value as non-editable text with onPress handler
 * - Editable mode: Full TextField with onValueChange for text input
 *
 * @param placeholder Placeholder text to display
 * @param value Current input value (null for clickable mode, non-null for editable)
 * @param onPress Press handler for expanding search (clickable mode)
 * @param onValueChange Text change callback (editable mode)
 * @param modifier Modifier for the search bar container
 */
@Composable
fun SearchBar(
    placeholder: String,
    value: String? = null,
    onPress: (() -> Unit)? = null,
    onValueChange: ((String) -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Theme tokens
    val backgroundColor = theme.colors.surface.default
    val textColor = theme.colors.onSurface.default
    val subtleColor = theme.colors.onSurface.default.copy(alpha = 0.6f)
    val cornerRadius = theme.radius.md
    val horizontalPadding = 16.dp
    val verticalPadding = 12.dp
    val iconSize = 20.dp
    val fontSize = 14.sp

    // Determine mode based on callback presence
    val isEditable = onValueChange != null

    // Container
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(cornerRadius),
        color = backgroundColor,
    ) {
        Row(
            modifier = Modifier
                .padding(horizontal = horizontalPadding, vertical = verticalPadding)
                .then(
                    if (!isEditable && onPress != null) {
                        Modifier.clickable(onClick = onPress)
                    } else {
                        Modifier
                    }
                ),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Search icon
            IconSymbol(
                name = "search",
                size = iconSize,
                color = subtleColor,
            )

            // Text field or clickable placeholder
            if (isEditable) {
                // Editable mode: TextField with input
                BasicTextField(
                    value = value ?: "",
                    onValueChange = onValueChange!!,
                    modifier = Modifier.weight(1f),
                    singleLine = true,
                    textStyle = TextStyle(
                        color = textColor,
                        fontSize = fontSize,
                    ),
                    cursorBrush = SolidColor(theme.colors.primary.default),
                    keyboardOptions = KeyboardOptions.Default.copy(
                        keyboardType = KeyboardType.Text
                    ),
                    decorationBox = { innerTextField ->
                        if (value.isNullOrEmpty()) {
                            Text(
                                text = placeholder,
                                color = subtleColor,
                                fontSize = fontSize,
                            )
                        }
                        innerTextField()
                    },
                )
            } else {
                // Clickable mode: Display placeholder or value as non-editable text
                Text(
                    text = value ?: placeholder,
                    color = if (value != null) textColor else subtleColor,
                    fontSize = fontSize,
                    modifier = Modifier.weight(1f),
                )
            }
        }
    }
}
